File: apps/dev-tools/src/modules/geo/climate.ts
// ──────────────────────────────────────────────────────────────
// Climate Simulation Module
// Temperature, wind, and precipitation from geology + latitude
// ──────────────────────────────────────────────────────────────

import { createNoise2D } from "simplex-noise";
import { ClimateConfig } from "./types";

export class ClimateSimulator {
    private windNoise: (x: number, y: number) => number;
    private moistureNoise: (x: number, y: number) => number;
    private tempVariation: (x: number, y: number) => number;

    constructor(
        private config: ClimateConfig,
        prng: () => number,
    ) {
        this.windNoise = createNoise2D(prng);
        this.moistureNoise = createNoise2D(prng);
        this.tempVariation = createNoise2D(prng);
    }

    /**
     * Calculate temperature at a point.
     * Factors: latitude, elevation, ocean proximity, local variation.
     * 
     * @param normalizedY - [0–1] Y position (0 = north pole, 1 = south pole)
     * @param elevation - [0–1] normalized elevation
     * @param oceanProximity - [0–1] how close to ocean (1 = at coast)
     * @param x, y - world coordinates for noise variation
     */
    getTemperature(
        normalizedY: number,
        elevation: number,
        oceanProximity: number,
        x: number,
        y: number,
    ): number {
        // Latitude gradient: hottest at equator (y ≈ 0.5), coldest at poles
        const latitudeFactor = 1 - 2 * Math.abs(normalizedY - 0.5); // 0 at poles, 1 at equator
        const baseTemp = this.config.globalMeanTemp +
            this.config.latitudeGradient * (latitudeFactor - 0.5);

        // Elevation lapse rate: -6.5°C per 1000m (assuming max elevation = 8848m)
        const elevationAboveSea = Math.max(0, elevation - 0.35) / 0.65; // normalize land elevation
        const elevationMeters = elevationAboveSea * 8848;
        const lapseRate = -6.5 * (elevationMeters / 1000);

        // Ocean moderation: coastal areas are milder
        const oceanModeration = oceanProximity * this.config.oceanWarmthFactor * 3;

        // Local random variation (±3°C)
        const localVariation = this.tempVariation(x / 500, y / 500) * 3;

        return baseTemp + lapseRate + oceanModeration + localVariation;
    }

    /**
     * Calculate precipitation / moisture availability.
     * Factors: base moisture noise, rain shadow from mountains, ocean proximity.
     * 
     * @param elevation - [0–1]
     * @param oceanProximity - [0–1]
     * @param windward - whether this point is on the windward side of mountains
     * @param x, y - world coordinates
     */
    getPrecipitation(
        elevation: number,
        oceanProximity: number,
        windward: boolean,
        x: number,
        y: number,
    ): number {
        // Base moisture from noise field
        const baseMoisture = (this.moistureNoise(x / 300, y / 300) + 1) / 2;

        // Ocean proximity increases moisture
        const oceanEffect = oceanProximity * 0.4;

        // Orographic effect: windward = wet, leeward = dry (rain shadow)
        const orographicEffect = windward
            ? Math.min(0.3, elevation * 0.5)   // uplift → rain
            : -Math.min(0.3, elevation * 0.4);  // rain shadow

        // Altitude reduction: very high = less precipitation
        const altitudeReduction = elevation > 0.7 ? -(elevation - 0.7) * 1.5 : 0;

        const result = (baseMoisture + oceanEffect + orographicEffect + altitudeReduction)
            * this.config.precipitationMultiplier;

        return Math.max(0, Math.min(1, result));
    }

    /**
     * Calculate wind exposure at a point.
     * Higher elevations and open terrain = more wind exposure.
     */
    getWindExposure(
        elevation: number,
        slopeAngle: number,
        x: number,
        y: number,
    ): number {
        const baseWind = (this.windNoise(x / 400, y / 400) + 1) / 2;
        const elevationBoost = elevation * 0.4;
        const slopeBoost = slopeAngle * 0.3;

        return Math.max(0, Math.min(1,
            baseWind * this.config.windStrength * 0.5 + elevationBoost + slopeBoost
        ));
    }

    /**
     * Determine if a point is on the windward side of terrain.
     * Simple approximation: compare elevation gradient against prevailing wind direction.
     */
    isWindward(
        elevationGradientX: number,
        elevationGradientY: number,
    ): boolean {
        const windDirRad = (this.config.prevailingWindDir * Math.PI) / 180;
        const windX = Math.sin(windDirRad);
        const windY = -Math.cos(windDirRad);

        // Dot product: positive means gradient faces the wind (windward)
        const dot = elevationGradientX * windX + elevationGradientY * windY;
        return dot > 0;
    }
}

File: packages/geo-core/src/climate.rs
use crate::geo_noise::NoiseManager;
use crate::types::ClimateConfig;
use noise::NoiseFn;

pub struct ClimateSimulator<'a> {
    config: &'a ClimateConfig,
    noise: &'a NoiseManager,
}

impl<'a> ClimateSimulator<'a> {
    pub fn new(config: &'a ClimateConfig, noise: &'a NoiseManager) -> Self {
        Self { config, noise }
    }

    pub fn get_temperature(
        &self,
        normalized_y: f64,
        elevation: f64,
        ocean_proximity: f64,
        wx: f64,
        wy: f64,
    ) -> f64 {
        let latitude_factor = 1.0 - 2.0 * (normalized_y - 0.5).abs();
        let base_temp =
            self.config.global_mean_temp + self.config.latitude_gradient * (latitude_factor - 0.5);

        let elevation_above_sea = (elevation - 0.35).max(0.0) / 0.65;
        let elevation_meters = elevation_above_sea * 8848.0;
        let lapse_rate = -6.5 * (elevation_meters / 1000.0);

        let ocean_moderation = ocean_proximity * self.config.ocean_warmth_factor * 3.0;
        let local_variation = self.noise.climate_noise.get([wx / 500.0, wy / 500.0]) * 3.0;

        base_temp + lapse_rate + ocean_moderation + local_variation
    }

    pub fn get_precipitation(
        &self,
        elevation: f64,
        ocean_proximity: f64,
        windward: bool,
        wx: f64,
        wy: f64,
    ) -> f64 {
        let base_moisture = (self.noise.climate_noise.get([wx / 300.0, wy / 300.0]) + 1.0) / 2.0;
        let ocean_effect = ocean_proximity * 0.4;

        let orographic_effect = if windward {
            (elevation * 0.5).min(0.3)
        } else {
            -(elevation * 0.4).min(0.3)
        };

        let altitude_reduction = if elevation > 0.7 {
            -(elevation - 0.7) * 1.5
        } else {
            0.0
        };

        let result = (base_moisture + ocean_effect + orographic_effect + altitude_reduction)
            * self.config.precipitation_multiplier;

        result.max(0.0).min(1.0)
    }

    pub fn get_wind_exposure(&self, elevation: f64, slope_angle: f64, wx: f64, wy: f64) -> f64 {
        let base_wind = (self.noise.climate_noise.get([wx / 400.0, wy / 400.0]) + 1.0) / 2.0;
        let elevation_boost = elevation * 0.4;
        let slope_boost = slope_angle * 0.3;

        (base_wind * self.config.wind_strength * 0.5 + elevation_boost + slope_boost)
            .max(0.0)
            .min(1.0)
    }

    pub fn is_windward(&self, grad_x: f64, grad_y: f64) -> bool {
        let wind_dir_rad = (self.config.prevailing_wind_dir * std::f64::consts::PI) / 180.0;
        let wind_x = wind_dir_rad.sin();
        let wind_y = -wind_dir_rad.cos();

        let dot = grad_x * wind_x + grad_y * wind_y;
        dot > 0.0
    }
}
